bfd: fill in PE load config directory entry.

jeremyd2019 opened the pull request:
This is filled in with the rva of _load_config_used if defined (much like _tls_used), and the size is the first 32-bit value at that symbol.

jeremyd2019 committed:
bfd: fill in PE load config directory entry.

This is filled in with the rva of _load_config_used if defined (much
like _tls_used), and the size is the first 32-bit value at that symbol.

Signed-off-by: Jeremy Drake <sourceware-bugzilla@jdrake.com>

lhmouse commented on bfd/peXXigen.c:
```diff
@@ -4573,6 +4575,43 @@ _bfd_XXi_final_link_postscript (bfd * abfd, struct coff_final_link_info *pfinfo)
+	  result = false;
+	}
+      /* the size is stored as the first 4 bytes at _load_config_used */
+      if (bfd_get_section_contents (abfd, h1->root.u.def.section->output_section,
```
I think it's possible to get these bytes directly form the input section. Looking at definition of `h1->root.u.def` (https://github.com/jeremyd2019/binutils-gdb/blob/master/include/bfdlink.h#L168-L173), I suspect `value` is the offset in that section, so this can be just 

```c
      if (h1->root.u.def.section != NULL
          && bfd_get_section_contents (abfd, h1->root.u.def.section, sz,
                                       h1->root.u.def.value, 4))
```

lhmouse replied:
> ```c
>       if (h1->root.u.def.section != NULL
>           && bfd_get_section_contents (abfd, h1->root.u.def.section, sz,
>                                        h1->root.u.def.value, 4))
> ```

This is not correct, because `h1->root.u.def.section` doesn't belong in `abfd`. In GDB if I `p abfd->filename` it gives the output EXE filename, so it's correct to use `h1->root.u.def.section->output_section`.

Issue here is that there should be proper checks for null pointers, and the `offset` value is incorrect.

cjacek commented:
Real-world load config use linker-provided CFG symbols. To avoid breaking the code (that is currently works because load config is ignored), I would suggest to define those missing symbols. It's find to define them as absolute 0 (at lest until ld supports CFG), see `__guard_*` symbols in lld-link for a complete list: https://github.com/llvm/llvm-project/blob/868c409d38fe01cc403bb2b93b716b119d2dde10/lld/COFF/Driver.cpp#L2454

lhmouse commented:
I debugged LD with GDB a little, and here's proposed correction for the if-statement beneath `/* the size is stored as the first 4 bytes at _load_config_used */`. I have renamed `char sz[4];`  to `char data[4];` as `sz` means 'size' elsewhere.

```c
      /* the size is stored as the first 4 bytes at _load_config_used */
      if (h1->root.u.def.section != NULL
	  && h1->root.u.def.section->output_section != NULL
	  && bfd_get_section_contents (abfd, h1->root.u.def.section->output_section, data,
				       h1->root.u.def.section->output_offset + h1->root.u.def.value, 4))
	pe_data (abfd)->pe_opthdr.DataDirectory[PE_LOAD_CONFIG_TABLE].Size =
	  bfd_get_32 (abfd, data);
```

### Explanation:

* `h1->root.u.def`: Definition of the symbol, `_load_config_used`.
* `h1->root.u.def.section`: Input sub-section from object files.
    For example, when `-fdata-sections` is passed to GCC, each individual variable is placed in its own section, so the size of this section is `sizeof(IMAGE_LOAD_CONFIG_DIRECTORY)`.
* `h1->root.u.def.value`: Offset of `_load_config_used` within its input sub-section.
    In the above example, the offset is zero.
* `abfd` and `h1->root.u.def.section->output_bfd`: BFD of the output.
    In this case it's the EXE file, which can be verified by printing `abfd->filename`.
* `h1->root.u.def.section->output_section`: Output section where input sub-section has been written.
* `h1->root.u.def.section->output_offset`: Offset of input sub-section in output section.

lhmouse commented:
Combined patch: https://github.com/lhmouse/MINGW-packages/blob/aaaa9ef039a2ac1c4c5061841fddc5a9954e2489/mingw-w64-binutils/9001-bfd-fill-in-PE-load-config-directory-entry.patch

jeremyd2019 commented:
> Real-world load config use linker-provided CFG symbols. To avoid breaking the code (that is currently works because load config is ignored), I would suggest to define those missing symbols. It's find to define them as absolute 0 (at lest until ld supports CFG), see `__guard_*` symbols in lld-link for a complete list: https://github.com/llvm/llvm-project/blob/868c409d38fe01cc403bb2b93b716b119d2dde10/lld/COFF/Driver.cpp#L2454

I'd rather not have to figure out how to do that 😉 .  The current mingw-w64 load config is actually all zeros already as long as mingw-w64 is not configured for CFG support.

lhmouse commented:
We can probably add a new option, namely `--dependent-load-flag <value>`, which overrides the `DependentLoadFlag` field.

It might be simple if it's just copying from
the `--image-base=`_org_ option.

jeremyd2019 commented:
> We can probably add a new option, namely `--dependent-load-flag <value>`, which overrides the `DependentLoadFlag` field.
> 
> It might be simple if it's just copying from the `--image-base=`_org_ option.



> I would suggest to define those missing symbols. It's find to define them as absolute 0 (at lest until ld supports CFG), see `__guard_*` symbols in lld-link for a complete list: https://github.com/llvm/llvm-project/blob/868c409d38fe01cc403bb2b93b716b119d2dde10/lld/COFF/Driver.cpp#L2454

These are both excellent ideas for follow-up patches :grin:

Reasons why it's not really that simple:
1. So far, all the symbols defined in pe.em/pep.em are things which are in `internal_extra_pe_aouthdr`  I don't know if this is meant to mirror actual PE headers, or if we can just add extra stuff to that struct.
2. in lld, they don't even use a symbol for dependent-load-flag, they just poke the value into the struct.  I don't know at what point in binutils link both the command-line options and the section contents/symbol table would be readily available to do that.

lhmouse commented:
> I'd rather not have to figure out how to do that 😉 . The current mingw-w64 load config is actually all zeros already as long as mingw-w64 is not configured for CFG support.

The symbols can be simply defined in the linker scripts, 'pe.sc' and 'pep.sc'. However I agree that it's probably not good to define zero or random constant values for now.

lhmouse commented:
there's usually a silence period before public patches can be checked in, so please remember to ping the patch after one week.

jeremyd2019 committed:
bfd: properly use bfd_get_symbol_leading_char in peXXigen.

This function returns the leading char to use, so we cannot just assume
it will always be '_' or '\0'.

After review, rationalize the error messages in
_bfd_XXi_final_link_postscript.  They now all correctly refer to
DataDirectory instead of DataDictionary, and use format strings so they
only need one translation.

Signed-off-by: Jeremy Drake <sourceware-bugzilla@jdrake.com>

jeremyd2019 committed:
bfd: add load config size workaround for i386 XP and earlier.

Per the Microsoft PE documentation, XP and earlier on i686 require the
Size field to be 64, rather than the actual size as required on other
architectures.  I have confirmed Windows 11 accepts either 64 or the
actual size for i386 images, but only the actual size for x86_64 images.

Signed-off-by: Jeremy Drake <sourceware-bugzilla@jdrake.com>

jeremyd2019 commented:
This was merged as https://github.com/jeremyd2019/binutils-gdb/compare/5de418ba70260e34d9246224805b958767d99482..fcdce142519644cdc359b3a7bc6eca6985e44c88
